state/upgrades_test.go

```diff
@@ -2155,42 +2155,167 @@ func (s *upgradesSuite) TestDeleteCloudImageMetadata(c *gc.C) {
+		m3.ModelUUID() + ":e": bson.M{"container-image-stream": "daily"},
+		"not-a-model":         bson.M{"other-setting": "val"},
+	}
+	for iter.Next(&rawSettings) {
+		expSettings := copyMap(rawSettings, nil)
+		delete(expSettings, "txn-queue")
+		delete(expSettings, "txn-revno")
+		delete(expSettings, "version")
+		id, ok := expSettings["_id"]
+		c.Assert(ok, jc.IsTrue)
+		idStr, ok := id.(string)
+		c.Assert(ok, jc.IsTrue)
+		c.Assert(idStr, gc.Not(gc.Equals), "")
+		if changes, ok := expectedChanges[idStr]; ok {
+			raw, ok := expSettings["settings"]
+			c.Assert(ok, jc.IsTrue)
+			settings, ok := raw.(bson.M)
+			c.Assert(ok, jc.IsTrue)
+			for k, v := range changes {
+				settings[k] = v
+			}
+		}
+		expectedSettings = append(expectedSettings, expSettings)
+	}
+	c.Assert(iter.Close(), jc.ErrorIsNil)
+
+	s.assertUpgradedData(c, UpgradeContainerImageStreamDefault,
+		expectUpgradedData{settingsColl, expectedSettings},
+	)
+}
+
+func (s *upgradesSuite) TestRemoveContainerImageStreamFromNonModelSettings(c *gc.C) {
+	// a model with a valid setting
+	m1 := s.makeModel(c, "m1", coretesting.Attrs{
+		"other-setting":          "val",
+		"container-image-stream": "released",
+	})
+	defer m1.Close()
+	// a model with a custom setting
+	m2 := s.makeModel(c, "m2", coretesting.Attrs{
+		"container-image-stream": "daily",
+		"other-setting":          "val",
+	})
+	defer m2.Close()
+
+	settingsColl, settingsCloser := s.state.db().GetRawCollection(settingsC)
+	defer settingsCloser()
+	// A document that isn't a model with an accidental setting
+	err := settingsColl.Insert(
 		bson.M{
-			"_id":      "bar",
-			"settings": bson.M{"container-image-stream": "", "other-setting": "val"},
+			"_id": "not-a-model",
+			"settings": bson.M{
+				"container-image-stream":                   "released",
+				"other-setting":                            "val",
+				unescapeReplacer.Replace("dotted.setting"): "value",
+				unescapeReplacer.Replace("dollar$setting"): "value",
+			},
 		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	// A document that doesn't have the setting
+	err = settingsColl.Insert(
 		bson.M{
-			"_id":      "baz",
-			"settings": bson.M{"container-image-stream": "daily"},
+			"_id": "applicationsetting",
+			"settings": bson.M{
+				"other-setting": "val",
+			},
 		},
 	)
 	c.Assert(err, jc.ErrorIsNil)
-
-	expectedSettings := []bson.M{
-		{
-			"_id":      "bar",
-			"settings": bson.M{"container-image-stream": "released", "other-setting": "val"},
-		},
-		{
-			"_id":      "baz",
-			"settings": bson.M{"container-image-stream": "daily"},
-		},
-		{
-			"_id":      "foo",
-			"settings": bson.M{"container-image-stream": "released", "other-setting": "val"},
+	// A document that has the setting, but it shouldn't be touched because it is a custom value.
+	err = settingsColl.Insert(
+		bson.M{
+			"_id": "otherapplication",
+			"settings": bson.M{
+				"container-image-stream": "custom",
```

Comment:
Should we be testing an encoded "." in the setting name for our removal code?

Reply:
this particular one is intentionally not touched, but I'm happy to sprinkle the custom "." across the tests to make sure we handle it correctly.
Would you like me to add a copyMap(..., escapeReplacer.Replace) to the original 2.3.6 upgrade code? And then remove the comment about not needing to, and adding tests that include '.' in the keys?
Or was it better that we caught all this by not doing that? :)